### ISSUE TITLE
take_until! does not play well with stream parsing

### ISSUE THREAD

**Issue by jpastuszek** [Geal/nom#397]
I was surprised to find out that `take_until!` returns Err if it fails to find the substring it is looking for.
Is there any variant that returns Incomplete instead?

**Comment by jpastuszek** [comment]
I have played with making all macros using `find_token`/`find_substring` function to return `IResult::Incomplete(Needed::Unknown)` if none was found. 
This fixes my stream parsers. I had to wrap some of this macros in `complete!` macro for parser that works on fixed size buffers though.

Looks like in nom it is far easier to convert `Incomplete` into `Error` rather than the way around.
Therefore I think that macros should by default work well with streams (return `Incomplete` rather than `Error` when more data is needed to finish matching) and require use of `complete!` macro when working with slices of fixed sizes.

`Error` should only happen when unexpected data was found and `Incomplete` (`Needed::Unknown`) when it could not find the data it was looking for.

My particular stream case was looking for `\n` terminated frames in stream (syslog):
```
named!(pub syslog_newline_frame<&[u8], &[u8]>, terminated!(take_until!("\n"), tag!(b"\n")));
```

And fixed buffer case now looks like this:
```
named!(timestamp<&[u8], DateTime<FixedOffset> >, return_error!(ErrorKind::Custom(2),
    complete!(terminated!(map_res!(take_until!(" "), parse::timestamp), tag!(b" ")))));
```

**Comment by RReverser** [comment]
@Geal I'm just new to `nom` (playing to see if it can replace our Ragel-based streaming parser in future), but this seems to be a blocking issue, got hit by `take_until_and_consume` when trying to use in a streaming fashion.

**Comment by Geal** [comment]
(just putting that reformulation of the problem here if I can't fix it today):
`take_until!` returns `Incomplete` if there's not at least the size of the substring it looks for (cf https://github.com/Geal/nom/blob/master/src/bytes.rs#L580 ), but there are cases where it won't work, like a `take_until("abcd")` where we get the chunks `123a` and `bcd`.
Possible angle here: look for the first byte of the substring, then use `tag!` from there, since `tag!` will handle partial input correctly.

**Comment by BurningMind** [comment]
I think it should also return `Incomplete` if we have `take_until!("f")` with chunk `abc` (unless you already meant that @Geal ).

**Comment by Geal** [comment]
Hello, this should be fixed by 484f672
